src/LinAlg/hiopLinSolverSparseCUSOLVER.cpp

```diff
@@ -0,0 +1,936 @@
+    for(int i = 0;i<n;++i) {
+
+      //go through EACH COLUMN OF L first
+      for(int j = Lp[i];j<Lp[i+1];++j) {
+        row = Li[j];
+        if(row != i){
+          //place (row, i) where it belongs!
+
+          mja_[mia_[row] + Mshifts[row]] = i;
+          Mshifts[row]++;
+        }
+      }
+      //each column of U next
+      for(int j = Up[i];j<Up[i+1];++j) {
+        row = Ui[j];
+        mja_[mia_[row] + Mshifts[row]] = i;
+        Mshifts[row]++;
+      }
+    }
+    free(Mshifts);
+    return 0;
+  }
+  void hiopLinSolverNonSymSparseCUSOLVER::firstCall()
+  {
+    assert(n_ == M_->n() && M_->n() == M_->m());
+    assert(n_>0);
+
+    // transfer triplet form to CSR form
+    // note that input is in lower triangular triplet form. First part is the sparse matrix, and the 2nd part are the additional dia_gonal elememts
+    // the 1st part is sorted by row
+    hiop::hiopMatrixSparseTriplet* M_triplet = dynamic_cast<hiop::hiopMatrixSparseTriplet*>(M_);
+
+    M_triplet->convertToCSR(nnz_, &kRowPtr_, &jCol_, &kVal_, &index_covert_CSR2Triplet_, &index_covert_extra_Diag2CSR_, extra_dia_g_nnz_map);
+
+    /*
+     * initialize cusolver parameters
+     */
+
+    //handles
+    cusparseCreate(&handle_); 
+    cusolverSpCreate(&handle_cusolver_);
+    cublasCreate(&handle_cublas_);
+
+    //descriptors
+    cusparseCreateMatDescr(&descr_A_);
+    cusparseSetMatType(descr_A_, CUSPARSE_MATRIX_TYPE_GENERAL);
+    cusparseSetMatIndexBase(descr_A_, CUSPARSE_INDEX_BASE_ZERO);
+
+    cusparseCreateMatDescr(&descr_M_);
+    cusparseSetMatType(descr_M_, CUSPARSE_MATRIX_TYPE_GENERAL);
+    cusparseSetMatIndexBase(descr_M_, CUSPARSE_INDEX_BASE_ZERO);
+
+    //info (data structure where factorization is stored)
+    cusolverSpCreateCsrluInfoHost(&info_lu_);
+    cusolverSpCreateGluInfo(&info_M_);
+
+    //KLU 
+
+    klu_defaults(&Common_) ;
+
+    //modify as you please
+    //KS: consider making a part of setup options that can be called from a user side
+
+    Common_.btf = 0;
+    Common_.ordering = 1;//COLAMD; use 0 for AMD
+    Common_.tol = 0.1;
+    Common_.scale = -1;
+    Common_.halt_if_singular=0;
+
+    free(mia_); 
+    free(mja_);
+    mia_ = NULL;
+    mja_ = NULL;
+    //allocate gpu data
+    cudaFree(devx_);
+    cudaFree(devr_);
+    devx_ = NULL;
+    devr_ = NULL;
+    checkCudaErrors(cudaMalloc(&devx_, n_ * sizeof(double)));
+    checkCudaErrors(cudaMalloc(&devr_, n_ * sizeof(double)));
+    this->newKLUfactorization();
+  }
+
+
+  int hiopLinSolverNonSymSparseCUSOLVER::newKLUfactorization()
+  {
+
+    Symbolic_ = klu_analyze(n_, kRowPtr_, jCol_, &Common_) ;
+
+    if(Symbolic_ == NULL){
+      //  klu_error_codes(Common_);
+      //throw error    
+      printf("symbolic NULL"); 
+    }
+
+    Numeric_ = klu_factor(kRowPtr_, jCol_, kVal_, Symbolic_, &Common_);
+    if(Numeric_ == NULL){
+      //klu_error_codes(Common_);
+      //throw error    
+      printf("numeric  NULL"); 
+    }
```

Comment:
Use HiOp logger instead of printing to stdout in order to maintain xSDK compliance:
```suggestion
    if(Symbolic_ == NULL){
      //  klu_error_codes(Common_);
      //throw error    
      nlp_->log->printf(hovError, "symbolic NULL"); 
    }

    Numeric_ = klu_factor(kRowPtr_, jCol_, kVal_, Symbolic_, &Common_);
    if(Numeric_ == NULL){
      //klu_error_codes(Common_);
      //throw error    
      nlp_->log->printf(hovError, "numeric  NULL"); 
    }
```

Also, need to use here proper error handler. These are both fatal errors.